pkg/controller/externalnode/controller.go

```diff
@@ -0,0 +1,386 @@
+}
+
+func (c *ExternalNodeController) enqueueExternalNodeUpdate(oldObj interface{}, newObj interface{}) {
+	en := newObj.(*v1alpha1.ExternalNode)
+	key, _ := keyFunc(en)
+	c.queue.Add(key)
+	klog.InfoS("Enqueued ExternalNode UPDATE event", "ExternalNode", klog.KObj(en))
+}
+
+func (c *ExternalNodeController) enqueueExternalNodeDelete(obj interface{}) {
+	en := obj.(*v1alpha1.ExternalNode)
+	key, _ := keyFunc(en)
+	c.queue.Add(key)
+	klog.InfoS("Enqueued ExternalNode DELETE event", "ExternalNode", klog.KObj(en))
+}
+
+// Run will create defaultWorkers workers (goroutines) which will process the ExternalEntity events from the work queue.
+func (c *ExternalNodeController) Run(stopCh <-chan struct{}) {
+	defer c.queue.ShutDown()
+
+	klog.InfoS("Starting", "controllerName", controllerName)
+	defer klog.InfoS("Shutting down", "controllerName", controllerName)
+
+	if !cache.WaitForNamedCacheSync(controllerName, stopCh, c.externalNodeListerSynced, c.externalEntityListerSynced) {
+		return
+	}
+	if err := c.reconcileExternalNode(); err != nil {
+		klog.ErrorS(err, "Failed to reconcile ExternalNodes")
+		return
+	}
+
+	for i := 0; i < defaultWorkers; i++ {
+		go wait.Until(c.worker, time.Second, stopCh)
+	}
+	<-stopCh
+}
+
+// reconcileExternalNode reconciles all the existing ExternalNodes and cleans up the stale ExternalEntities.
+func (c *ExternalNodeController) reconcileExternalNode() error {
+	externalNodes, err := c.externalNodeLister.List(labels.Everything())
+	if err != nil {
+		return err
+	}
+	enUIDEENameMap := make(map[types.UID]string)
+	for _, en := range externalNodes {
+		if err = c.addExternalNode(en); err != nil {
+			return err
+		}
+		eeName := genExternalEntityName(en)
+		enUIDEENameMap[en.UID] = eeName
+	}
+	externalEntities, err := c.externalEntityLister.List(labels.Everything())
+	if err != nil {
+		return err
+	}
+	for _, ee := range externalEntities {
+		if (len(ee.OwnerReferences) > 0) && (ee.OwnerReferences[0].Kind == "ExternalNode") {
+			// Clean up stale ExternalEntities when ExternalNode no longer exists or
+			// when interface[0] name is changed.
+			if eeName, ok := enUIDEENameMap[ee.OwnerReferences[0].UID]; !ok || (ok && (eeName != ee.Name)) {
+				err = c.crdClient.CrdV1alpha2().ExternalEntities(ee.Namespace).Delete(context.TODO(), ee.Name, metav1.DeleteOptions{})
+				if err != nil {
+					return err
+				}
+			}
+		}
+	}
+	return nil
+}
+
+// worker is a long-running function that will continually call the processNextWorkItem function in
+// order to read and process a message on the work queue.
+func (c *ExternalNodeController) worker() {
+	for c.processNextWorkItem() {
+	}
+}
+
+func (c *ExternalNodeController) processNextWorkItem() bool {
+	obj, quit := c.queue.Get()
+	if quit {
+		return false
+	}
+	defer c.queue.Done(obj)
+
+	if key, ok := obj.(string); !ok {
+		c.queue.Forget(obj)
+		klog.Errorf("Expected string in ExternalNode work queue but got %#v", obj)
```

Comment:
Change to ErrorS

Reply:
There is no err generated here, so I guess we can not use ErrorS in this case?

Reply:
I think we just use InfoS() in this case. But I am fine with Errorf() too.